### PR TITLE
Feedback Backend Integration

### LINKED ISSUE #5073: Create backend endpoint and model for feedback submissions
## Background
To support our user feedback initiative, TDP needs a backend endpoint to receive general application feedback submitted through the frontend. This endpoint will accept both anonymous and authenticated submissions and store the data in a new Django model designed for analysis and reporting.

The model should meet the following requirements:
- **Nullable relationship to `User`** to support both anonymized and logged-in feedback
- **Enumerated category field** to map to Likert scale or other rating types
- **Unbounded text field** to store optional user comments (consider adding a soft character limit for DB hygiene)
- **Status field** to indicate whether feedback has been _viewed_, _responded to_, or _resolved_

This infrastructure lays the foundation for real-time insight collection and user-driven product improvement.

### Why It Matters
Capturing general feedback directly within the application helps identify usability issues, gather sentiment, and support continuous improvement efforts. By centralizing and tagging feedback, we also make it easier for OFA, UX, and PM teams to analyze trends, validate hypotheses, and prioritize enhancements. This supports our **Continuous Feedback Loops (CFL)** initiative and user trust goals.

### What’s Changing
- Backend API endpoint added for general feedback submissions
- New Django model created to store, tag, and track feedback
- Option to associate feedback with a logged-in user or store it anonymously
- Unit tests ensure the endpoint and model handle valid and edge-case inputs

### Acceptance Criteria
- [ ] Django model for general feedback is created with required fields and migrated successfully  
- [ ] DRF endpoint is available to POST feedback from the frontend  
- [ ] Submissions are stored with correct metadata (metadata documented in mockup) 
- [ ] Input validation and error handling are implemented  
- [ ] Unit tests are written for both model and endpoint (including edge cases)  
- [ ] Endpoint is integrated with frontend UI (if applicable at time of development)  
- [ ] Full Testing Checklist is completed  
- [ ] README is updated if implementation details or API usage changes  
- [ ] [Release digest](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/JSIbkxujKUxhKv2cXFEg/) is updated with backend testing steps and example payload

### Usability Testing Criteria
_PM/UX/OFA Only:_
- [ ] Feedback submitted anonymously and with user accounts is stored correctly  
- [ ] Categories align with frontend input types (e.g., Likert scale)  
- [ ] Internal teams can view and interpret collected feedback from the admin dashboard or DB queries

### Notes
- This endpoint may be reused for contextual feedback in future iterations  
- Consider defaulting `status` to `unread` and adding a management command or admin view to triage feedback  
- Could eventually pipe to Sentry, Slack, or a dashboard for faster response workflows

### Supporting Documentation 
- [UX Mockups](https://www.figma.com/design/irgQPLTrajxCXNiYBTEnMV/TDP-Mockups-For-Feedback?node-id=11639-10947&t=agmruKjJX3ftMZkr-0)

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Add admin interface to individually and bulk acknowledge feedback
- Add mixin class for the incoming admin review capability for change requests and this work
- Add admin template for acking feedback
- Add migration
- Added api endpoints to create, list, and get feedback models
- Added tests
Pull request closes #5073 

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```
Perform the following steps as authenticated, un-authenticated users.

1. Open http://localhost:3000/v1/feedback
2. Verify the list of objects returned match your role/user info
3. Create a new feedback model
4. Verify in the admin console the model was created and hasn't been `acked`
5. From the admin console test "acking" individual feedback and bulk "acking" feedback via actions
6. Verify after "acking" the feedback that the `reviewed_at` and `reviewed_by` are updated appropriately and that you cannot "re-ack" feedback.
7. Notice the UI difference between an "acked" and "un-acked" feedback in the admin console form view

To submit feeback anonymously, you can use the following command:
```
curl -X POST \
  http://localhost:8080/v1/feedback/ \
  -H 'Content-Type: application/json' \
  -d '{
    "rating": 5,
    "feedback": "Anonymous feedback test"
}'
```

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
### Acceptance Criteria
- [ ] Django model for general feedback is created with required fields and migrated successfully  
- [ ] DRF endpoint is available to POST feedback from the frontend  
- [ ] Submissions are stored with correct metadata (metadata documented in mockup) 
- [ ] Input validation and error handling are implemented  
- [ ] Unit tests are written for both model and endpoint (including edge cases)  
- [ ] Endpoint is integrated with frontend UI (if applicable at time of development)  
- [ ] Full Testing Checklist is completed  
- [ ] README is updated if implementation details or API usage changes  
- [ ] [Release digest](https://app.gitbook.com/o/Yd4Wv0Fi89kSKpmakWrn/s/JSIbkxujKUxhKv2cXFEg/) is updated with backend testing steps and example payload

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/users/views.py [hunk 1]
```diff
@@ -88,3 +89,40 @@ class GroupViewSet(viewsets.GenericViewSet, mixins.ListModelMixin):
...
+    """Feedback viewset."""
+
+    queryset = Feedback.objects.all()
+    serializer_class = FeedbackSerializer
+    permission_classes = ()
+
+    def retrieve(self, request, pk=None):
+        """Return a specific feedback."""
+        item = get_object_or_404(self.queryset, pk=pk)
+        if request.user.is_authenticated and request.user == item.user:
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/views.py]: should an admin be able to view any individual item, or just ones they've submitted?

also not strictly necessary since i don't think we'd use it anywhere else for now, but i'd love an `IsObjectOwner` permission class that permits viewing all for admins and only owned objects (`request.user == item.user`) for other roles. that might help DRY some of the checks here and in `list`

File: tdrs-backend/tdpservice/users/serializers.py [hunk 2]
```diff
@@ -180,3 +181,28 @@ def update(self, instance, validated_data):
...
+class FeedbackSerializer(serializers.ModelSerializer):
+    """Serializer for user feedback."""
+
+    class Meta:
+        """Serializer metadata."""
+
+        model = Feedback
+        fields = (
+            'id',
+            'user',
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/serializers.py]: can `user` be a read only field as well? as an anonymous user, it currently lets me pick a user to submit feedback as 
image

it should be possible to write the user as part of the model creation without allowing it as a post body parameter

Comment (jtimpe) [reply tdrs-backend/tdpservice/users/serializers.py]: actually, unless a user specifically selects themself in the feedback form, the user isn't assigned to the request user at all. I understand this is to allow anonymous users, but i think anonymous users can create feedback, as well as the `anonymous` form submission field can be used to control whether the authenticated user is attached to the request. but the api should not allow user assignments. related to my other comment as well

**Review by jtimpe — CHANGES_REQUESTED** [review]
Looking good overall, just have a suggestion on how to handle the `user` aspect.

File: tdrs-backend/tdpservice/users/serializers.py (hunk 2 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/users/serializers.py]: Made user read only and removed it from the serializers response. We now the get the user from the request object in the backend from the middleware.

File: tdrs-backend/tdpservice/users/views.py (hunk 1 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/users/views.py]: Added new permissions class for this. Thank you for the suggestion!

**Review by lhawkman27 — APPROVED** [review]
Tested it out and looks like it's working the way it should.

File: tdrs-backend/tdpservice/users/views.py [hunk 3]
```diff
@@ -88,3 +95,36 @@ class GroupViewSet(viewsets.GenericViewSet, mixins.ListModelMixin):
...
+        try:
+            feedback_id = response.data['id']
+            feedback = Feedback.objects.get(id=feedback_id)
+
+            # Force anonymity if user is None to prevent us from know if authenticated users chose to remain anonymous
+            if request.user is None or isinstance(request.user, AnonymousUser):
+                feedback.anonymous = True
+
+            if not feedback.anonymous:
+                feedback.user = request.user
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/users/views.py:125]: nice :)

**Review by jtimpe — APPROVED** [review]

**Review by raftmsohani — APPROVED** [review]
